ecChronos will break if repair interval is shorter than the initial delay

**Description:**
By setting the repair interval to something shorter than one day and not setting the initial delay for a table in the ecc.yml, ecChronos will fail starting. Initial delay must always be shorter than the repair interval as it is now.

**Detailed description:**
Setting repair interval to, for example, 30 seconds, but keep the default value for initial delay (which is 24 hours) by not specifying any in the ecc.yml will result in an exception and ecChronos will fail to start. This happens since version 5.0.3 of ecChronos.

The workaround for this is to explicitly set the initial_delay to a shorter value and not relying on the default value. 

Exception
[main] [org.springframework.boot.SpringApplication] Application run failed org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'ECChronos' defined in URL [jar:file:/usr/share/ecchronos/lib/application-5.0.3.jar!/com/ericsson/bss/cassandra/ecchronos/application/spring/ECChronos.class]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate
...
Caused by: java.lang.IllegalArgumentException: Schedule \\"ks\\".\\"tb1\\" initial delay must be shorter than the repair interval. Repair interval: 30 seconds, initial delay : 86400 seconds
...

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,6 +1,8 @@
 # Changes
 ## Version 5.0.4
 
+* ecChronos will break if repair interval is shorter than the initial delay - Issue #667
+
 ## Version 5.0.3
 
 * Spring Framework URL Parsing with Host Validation is vulnerable - Issue #661

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/Interval.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/Interval.java
@@ -41,9 +41,21 @@ public class Interval
     }
 
     @JsonProperty("time")
+    public final long getTime()
+    {
+        return myTime;
+    }
+
+    @JsonProperty("time")
     public final void setTime(final long time)
     {
         myTime = time;
+    }
+
+    @JsonProperty("unit")
+    public final TimeUnit getUnit()
+    {
+        return myUnit;
     }
 
     @JsonProperty("unit")

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/repair/RepairConfig.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/repair/RepairConfig.java
@@ -22,9 +22,13 @@ import com.ericsson.bss.cassandra.ecchronos.core.repair.RepairConfiguration;
 import com.ericsson.bss.cassandra.ecchronos.core.repair.RepairOptions;
 import com.ericsson.bss.cassandra.ecchronos.core.utils.UnitConverter;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class RepairConfig
 {
+    private static final Logger LOG = LoggerFactory.getLogger(RepairConfig.class);
+
     private static final int DAYS_INTERVAL = 7;
     private static final int DAYS_WARNING = 8;
     private static final int DAYS_ERROR = 10;
@@ -53,6 +57,12 @@ public class RepairConfig
     public final void setPriority(final Priority priority)
     {
         myPriority = priority;
+    }
+
+    @JsonProperty("interval")
+    public final Interval getRepairInterval()
+    {
+        return myRepairInterval;
     }
 
     @JsonProperty("interval")
@@ -140,13 +150,13 @@ public class RepairConfig
                     repairIntervalSeconds, warningIntervalSeconds));
         }
 
-        if (initialDelaySeconds >= repairIntervalSeconds)
+        if (repairIntervalSeconds < initialDelaySeconds)
         {
-            throw new IllegalArgumentException(String.format(
-                        "%s initial delay must be shorter than the repair interval."
-                        + " Repair interval: %d seconds, initial delay : %d seconds", repairConfigType,
-                repairIntervalSeconds, initialDelaySeconds));
+            LOG.warn("{} repair interval ({}s) is shorter than initial delay ({}s). Will use {}s as initial delay.",
+                    repairConfigType, repairIntervalSeconds, initialDelaySeconds, repairIntervalSeconds);
+            myInitialDelay = new Interval(myRepairInterval.getTime(), myRepairInterval.getUnit());
         }
+
         long errorIntervalSeconds = myAlarm.getErrorInterval().getInterval(TimeUnit.SECONDS);
         if (warningIntervalSeconds >= errorIntervalSeconds)
         {

--- a/application/src/main/resources/ecc.yml
+++ b/application/src/main/resources/ecc.yml
@@ -97,8 +97,8 @@ repair:
     time: 7
     unit: days
   ##
-  ## Initial delay for new tables. New tables are always assumed to have been repaired in the past, however a delay
-  ## can be set for the first repair. This will not affect subsequent repairs and defaults to one day.
+  ## Initial delay for new tables. New tables are always assumed to have been repaired in the past by the interval.
+  ## However, a delay can be set for the first repair. This will not affect subsequent repairs and defaults to one day.
   ##
   initial_delay:
     time: 1

--- a/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java
+++ b/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java
@@ -196,7 +196,13 @@ public class TestConfig
         File file = new File(classLoader.getResource("initial_delay_greater_than_repair.yml").getFile());
 
         ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory());
-        assertThatExceptionOfType(JsonMappingException.class).isThrownBy(() -> objectMapper.readValue(file, Config.class));
+        Config config = objectMapper.readValue(file, Config.class);
+        GlobalRepairConfig repair = config.getRepairConfig();
+        Interval repint = repair.getRepairInterval();
+        Interval intdel = repair.getInitialDelay();
+
+        assertThat(repint.getTime()).isEqualTo(intdel.getTime());
+        assertThat(repint.getUnit()).isEqualTo(intdel.getUnit());
     }
 
     @Test

--- a/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/state/RepairStateImpl.java
+++ b/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/state/RepairStateImpl.java
@@ -207,14 +207,12 @@ public class RepairStateImpl implements RepairState
         long runIntervalInMs = myRepairConfiguration.getRepairIntervalInMs();
         long initialDelayInMs = myRepairConfiguration.getInitialDelayInMs();
         long assumedRepairedAt = System.currentTimeMillis() - runIntervalInMs + initialDelayInMs;
-
         if (LOG.isInfoEnabled())
         {
             LOG.info("Assuming the table {} is new. Next repair will occur at {}.",
                     myTableReference,
                     MY_DATE_FORMAT.get().format(new Date(assumedRepairedAt + runIntervalInMs)));
         }
-        
         return assumedRepairedAt;
     }
 

--- a/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/state/RepairStateImpl.java
+++ b/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/state/RepairStateImpl.java
@@ -208,10 +208,13 @@ public class RepairStateImpl implements RepairState
         long initialDelayInMs = myRepairConfiguration.getInitialDelayInMs();
         long assumedRepairedAt = System.currentTimeMillis() - runIntervalInMs + initialDelayInMs;
 
-        LOG.info("Assuming the table {} is new. Next repair will occur at {}.",
-                myTableReference,
-                MY_DATE_FORMAT.get().format(new Date(assumedRepairedAt + runIntervalInMs)));
-
+        if (LOG.isInfoEnabled())
+        {
+            LOG.info("Assuming the table {} is new. Next repair will occur at {}.",
+                    myTableReference,
+                    MY_DATE_FORMAT.get().format(new Date(assumedRepairedAt + runIntervalInMs)));
+        }
+        
         return assumedRepairedAt;
     }
 

--- a/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/state/RepairStateImpl.java
+++ b/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/state/RepairStateImpl.java
@@ -208,11 +208,9 @@ public class RepairStateImpl implements RepairState
         long initialDelayInMs = myRepairConfiguration.getInitialDelayInMs();
         long assumedRepairedAt = System.currentTimeMillis() - runIntervalInMs + initialDelayInMs;
 
-        if (LOG.isInfoEnabled())
-        {
-            LOG.info("Assuming the table {} is new. Next repair will occur at {}",
-                    myTableReference, MY_DATE_FORMAT.get().format(new Date(assumedRepairedAt + runIntervalInMs)));
-        }
+        LOG.info("Assuming the table {} is new. Next repair will occur at {}.",
+                myTableReference,
+                MY_DATE_FORMAT.get().format(new Date(assumedRepairedAt + runIntervalInMs)));
 
         return assumedRepairedAt;
     }

--- a/docs/GETTING_STARTED.md
+++ b/docs/GETTING_STARTED.md
@@ -31,11 +31,11 @@ Installation information can be found in [SETUP.md](SETUP.md).
 
 When ecChronos is started, it will read repair history to see the status of the repairs.
 If repair history exists for a table, ecChronos will use the last time repair was run and add the repair interval to calculate the next repair time. 
-On the other hand, if there's no repair history present for a table, the ecChronos will assume that the table has been repaired in the past using 
-initial delay to calculate.
+On the other hand, if there's no repair history present for a table, ecChronos will assume the table has been repaired in the past, but an initial delay (defaults to one day) will then offset the first repair.
+If the repair interval is shorter than the configured initial delay, initial delay will be set to the same value as the repair interval.
 
 The assumption is done in the following way:
-* Initial delay is configurable (default 1 day)
+* Initial delay = min(initial delay, repair interval)
 * Completed at = (start time - repair interval + initial delay)
 * Next repair = (completed at + repair interval)
 

--- a/docs/autogenerated/EccYamlFile.md
+++ b/docs/autogenerated/EccYamlFile.md
@@ -83,8 +83,8 @@
 * time: 7
 * unit: days
 #
-# Initial delay for new tables. New tables are always assumed to have been repaired in the past, however a delay
-#  can be set for the first repair. This will not affect subsequent repairs and defaults to one day.
+# Initial delay for new tables. New tables are always assumed to have been repaired in the past by the interval.
+# However, a delay can be set for the first repair. This will not affect subsequent repairs and defaults to one day.
 #
 **initial_delay:**
 * time: 1

